lib/internal/bootstrap/loaders.js

```diff
@@ -230,61 +235,88 @@ const getOwn = (target, property, receiver) => {
+    'internal/modules/esm/create_dynamic_module');
+  const url = this.getURL();
+  return this.esmFacade = createDynamicModule(
+    [], [...this.exportKeys, 'default'], url, (reflect) => {
+      this.reflect = reflect;
+      this.syncExports();
+      reflect.exports.default.set(this.exports);
+    });
+};
+
 // Provide named exports for all builtin libraries so that the libraries
 // may be imported in a nicer way for ESM users. The default export is left
 // as the entire namespace (module.exports) and wrapped in a proxy such
 // that APMs and other behavior are still left intact.
-NativeModule.prototype.proxifyExports = function() {
-  this.exportKeys = Object.keys(this.exports);
-
-  const update = (property, value) => {
-    if (this.reflect !== undefined &&
-        ObjectPrototype.hasOwnProperty(this.reflect.exports, property))
-      this.reflect.exports[property].set(value);
-  };
-
-  const handler = {
-    __proto__: null,
-    defineProperty: (target, prop, descriptor) => {
-      // Use `Object.defineProperty` instead of `Reflect.defineProperty`
-      // to throw the appropriate error if something goes wrong.
-      Object.defineProperty(target, prop, descriptor);
-      if (typeof descriptor.get === 'function' &&
-          !Reflect.has(handler, 'get')) {
-        handler.get = (target, prop, receiver) => {
-          const value = Reflect.get(target, prop, receiver);
-          if (ObjectPrototype.hasOwnProperty(target, prop))
-            update(prop, value);
-          return value;
-        };
-      }
-      update(prop, getOwn(target, prop));
-      return true;
-    },
-    deleteProperty: (target, prop) => {
-      if (Reflect.deleteProperty(target, prop)) {
-        update(prop, undefined);
-        return true;
-      }
-      return false;
-    },
-    set: (target, prop, value, receiver) => {
-      const descriptor = Reflect.getOwnPropertyDescriptor(target, prop);
-      if (Reflect.set(target, prop, value, receiver)) {
-        if (descriptor && typeof descriptor.set === 'function') {
-          for (const key of this.exportKeys) {
-            update(key, getOwn(target, key, receiver));
-          }
-        } else {
-          update(prop, getOwn(target, prop, receiver));
-        }
-        return true;
-      }
-      return false;
+NativeModule.prototype.syncExports = function() {
+
+  // const update = (property, value) => {
+  //   if (this.reflect !== undefined &&
+  //       ObjectPrototype.hasOwnProperty(this.reflect.exports, property))
+  //     this.reflect.exports[property].set(value);
+  // };
+
+  const names = this.exportKeys;
+  if (this.reflect) {
+    for (let i = 0; i < names.length; i++) {
+      const exportName = names[i];
+      if (exportName === 'default') continue;
+      this.reflect.exports[exportName].set(
+        getOwn(this.exports, exportName, this.exports)
+      );
     }
-  };
+  }
 
-  this.exports = new Proxy(this.exports, handler);
+  // const handler = {
```

Comment:
I assume this code will be removed before shipping?

Reply:
yes, cleanup and docs are pending semantic reviews

Reply:
Sounds good, I'll leave this thread as a reminder.

Reply:
@jkrems should be good for another review